### ISSUE TITLE
Upsampling/Oversampling for PhaseScope/PrettyScope

### ISSUE THREAD

**Issue by elanhickler** [RobinSchmidt/RS-MET#57]
No idea what this feature involves, but you said you could add a feature to remove unrealistic jagged lines, sharp corners, etc. As seen here: https://www.youtube.com/watch?v=SvCfz-TY6Go

I will gladly pay for your time on this, I'd like to have it as soon as possible, assuming it's not a big investment.

**Comment by RobinSchmidt** [comment]
i was thinking about connecting the incoming signal "dots" not with straight lines but cubic spline segments. it's a bit of math to work out, i'm not sure, how hard this will be, maybe a few hours. however, i'm not sure how you would draw splines in OpenGL - i can do it straightforwardly in PhaseScope because i do the line-drawing myself anyway (i would just replace the dotted-line-drawing algorithm with a dotted-spline-drawing algorithm). but in OpenGL, the lowest level you can access is the line-drawer (i think)....unless you also draw lines with dots which will be expensive. so, i'm not sure, how easily it will translate. ....maybe for PrettyScope, straightforward oversampling with lowpass filtering might be the better approach

**Comment by RobinSchmidt** [comment]
>  i guess, i'll have to do a line integral for this

something like that:

http://tutorial.math.lamar.edu/Classes/CalcIII/LineIntegralsPtI.aspx

but that's really the 2nd step. first i need to have parametric formulas for the spline segment itself. i guess, i can just use the regular cubic spline interpolation formulas for x and y separately. ...hopefully

**Comment by RobinSchmidt** [comment]
using an oversampling approach would just mean to compute more intermediate data points and them connecting *them* with line segments. this would also smooth the curve, but is arguably less elegant. ...but might be the only way to do it in OpenGL anyway

**Comment by RobinSchmidt** [comment]
ah - no - this is even more relevant:
http://tutorial.math.lamar.edu/Classes/CalcII/ParaArcLength.aspx
the link above is for more general stuff, this is a simpler special case which applies here

**Comment by RobinSchmidt** [comment]
i think, it will eventually boil down to such an integral:

http://www.wolframalpha.com/input/?i=integral+sqrt(a_0+%2B+a_1+t+%2B+a_2+t%5E2+%2B+a_3+t%5E3+%2B+a_4+t%5E4)+dt+from+0+to+1

because the derivative of a cubic is a quadratic, squaring it yields a quartic - so under the square root we'll get a sum of two quartics which is still a quartic. unfortunately, it exceeds standard computation time of wolfram alpha. ...but i'm really trying to do the second step before the first here. so, nevermind

**Comment by elanhickler** [comment]
I stole this from dood.al if it helps:

```javascript

	generateSmoothedSamples : function (oldSamples, samples, smoothedSamples)
	{
		//this.createLanczosKernel();
		var bufferSize = this.bufferSize;
		var allSamples = this.allSamples;
		var nSmoothedSamples = this.nSmoothedSamples;
		var a = this.a;
		var steps = this.steps;
		var K = this.K;

		for (var i=0; i<bufferSize; i++)
		{
			allSamples[i] = oldSamples[i];
			allSamples[bufferSize+i] = samples[i];
		}

		/*for (var s= -a+1; s<a; s++)
		{
			for (var r=0; r<steps; r++)
			{
				if (r==0 && !(s==0)) continue;
				var kernelPosition = -r+s*steps;
				if (kernelPosition<0) k = K[-kernelPosition];
				else k = K[kernelPosition];

				var i = r;
				var pStart = bufferSize - 2*a + s;
				var pEnd = pStart + bufferSize;
				for (var p=pStart; p<pEnd; p++)
				{
					smoothedSamples[i] += k * allSamples[p];
					i += steps;
				}
			}
		}*/

		var pStart = bufferSize - 2*a;
		var pEnd = pStart + bufferSize;
		var i = 0;
		for (var position=pStart; position<pEnd; position++)
		{
			smoothedSamples[i] = allSamples[position];
			i += 1;
			for (var r=1; r<steps; r++)
			{
				var smoothedSample = 0;
				for (var s= -a+1; s<a; s++)
				{
					var sample = allSamples[position+s];
					var kernelPosition = -r+s*steps;
					if (kernelPosition<0) smoothedSample += sample * K[-kernelPosition];
					else smoothedSample += sample * K[kernelPosition];
				}
				smoothedSamples[i] = smoothedSample;
				i += 1;
			}
		}

		smoothedSamples[nSmoothedSamples-1] = allSamples[2*bufferSize-2*a];
	},

    createLanczosKernel : function ()
    {
    	this.K = new Float32Array(this.radius);
    	this.K[0] = 1;
    	for (var i =1; i<this.radius; i++)
    	{
    		var piX = (Math.PI * i) / this.steps;
    		var sinc = Math.sin(piX)/piX;
    		var window = this.a * Math.sin(piX/this.a) / piX;
    		this.K[i] = sinc*Math.pow(window, this.lanczosTweak);
    	}
    }
}
```

**Comment by RobinSchmidt** [comment]
are you suggesting to use this as upsampling filter?

**Comment by RobinSchmidt** [comment]
if we do oversampling, i'd probably rather just throw in a bessel filter

**Comment by elanhickler** [comment]
No, I am simply letting you know what the code is behind dood.al, because I wish PrettyScope looked exactly like it (minus a few things, plus I need lots more features), but use whatever algorithm you want, as long as it works!

**Comment by RobinSchmidt** [comment]
ok, we'll see. i'm offline now until sunday

**Comment by elanhickler** [comment]
Just want to ask about something. I added mouse control to the canvas of PrettyScope essentially turning it into an X/Y pad that can also output your mouse movements as audio. However, it doesn't looks as nice as I'd like, because the mouse is only updated so often and you get little corners despite smoothing.

If/when you implement this oversampling thing for PrettyScope, will it be relevant to smooth 2d mouse movements? Maybe some extreme smoothing setting (since mouse movements and audio rate movements are very different speeds).

**Comment by RobinSchmidt** [comment]
not only do they have a different speed/rate but also are mouse movements potentially non-equidistant which pretty much rules out any common filtering technique for upsampling the mouse events. hmmm...actually, when you apply smoothing to the shiftX/Y parameters, what you do is to sample-and-hold the incoming values and then apply a 1st order lowpass to that S-H-signal. you could perhaps replace that 1st order lowpass by a 2nd order one to get rid of corners.

**Comment by RobinSchmidt** [comment]
well, if your use rosic::ExponentialSmoother for smoothing, this class actually encapsulates the sample-and-hold process and the 1st order lowpass into one convenient object. you could simply put another lowpass after it, for example, a rosic::OnePoleFilter.

i could even further encapsulate that into a class myself, like rosic::SecondOrderSmoother or something

**Comment by elanhickler** [comment]
What's the difference between onepole and exponentialsmoother?

**Comment by elanhickler** [comment]
I need a "get set target value" / "get current value" function. OnePole filter only has getSample.

**Comment by elanhickler** [comment]
![image](https://user-images.githubusercontent.com/9116575/29996768-6eb8f696-8fb9-11e7-9aa7-b40a49c5e375.png)

**Comment by RobinSchmidt** [comment]
smoother2 doesn't need a target value. it always gets an input that has the S/H already applied. think of it as a 3 step process: 

1: sample-and-hold
2: lowpass 1
3: lowpass 2 

step 1 and 2 are alamgamated into ExponentialSmoother. lowpass 2 simply takes the output of the 2nd. it doesn't have its own target value.

maybe i should simply implement it myself in rosic

**Comment by elanhickler** [comment]
noooooooooooo, I think I want to try higher poles.

**Comment by RobinSchmidt** [comment]
hmm...maybe a general higher order Smoother class could be written that has a setOrder() function. no need to restrict yourself to powers of two, btw.

**Comment by RobinSchmidt** [comment]
conceptually, it's just sample-and-hold and then apply any number of 1st order lowpasses one after another. of course, some optimizations are possible: if all lowpasses have the same settings, you don't need to have them all store their own cutoff, sample-rate etc. or even coeffs. ...and perhaps one would want to auto-scale the actual time-constant with the order in some way

**Comment by RobinSchmidt** [comment]
...also, it's not necessary that all lowpasses have the same cutoff. it's not even necessarry to use a bunch of 1st order lowpasses - instead a single high-order lowpass could be used. perhaps one that approximates gaussian impulse response. together with sample-and-hold, the transition would approximate a nice sigmoid (the integral of a gaussian bell curve)

**Comment by elanhickler** [comment]
Bah, just tried fourpolefilter. It's all not good. What I envision is more of an upsampling filter or a realtime bezier system / filter... like... your onepolefilter has an internal sample & hold, I guess instead of using a onepole you would use a system of a history of data points that you interpolate between.

**Comment by RobinSchmidt** [comment]
...but that might be too fancy. i guess, an auto-scaled series of equal lowpasses should be good enough and more straightforward and efficient

**Comment by RobinSchmidt** [comment]
> Bah, just tried fourpolefilter

what kind of four pole filter?

**Comment by elanhickler** [comment]
your FourPoleFilter.

Is there a steep slope filter class I could try?

**Comment by RobinSchmidt** [comment]
> instead of using a onepole you would use a system of a history of data points that you interpolate between.

yeah, the problem is that such a thing is difficult to do in realtime, because you might want to use future datapoints now that you don't have yet. for example, to connect two points with a cubic spline, i need to define the positions of the points and the derivatives there. but to assign meaningful values to the derivatives of the current datapoint, i may want to to know the next one (to use the difference of next-current for the derivative)

**Comment by RobinSchmidt** [comment]
> your FourPoleFilter.

that's a bad one. it's two biquads and will produce overshoot ...right?

**Comment by RobinSchmidt** [comment]
for time domain work, the best would probably be a gaussian. but a series of equal 1st order filters may be not that bad either

**Comment by elanhickler** [comment]
overshoot? i dunno, I'm just still getting corners. Oh, wouldn't a high order slope cause ringing/overshooting anyway?

**Comment by RobinSchmidt** [comment]
> wouldn't a high order slope cause ringing/overshooting anyway?

that depends on the response type of the filter. butterworth: certainly. elliptic: even more. bessel: much less so. gaussian: probably nicest shape.

**Comment by RobinSchmidt** [comment]
> Is there a steep slope filter class I could try?

you could try EngineersFilter with the Bessel response. ..i'd like to add gaussian someday but it's not there yet

**Comment by elanhickler** [comment]
what is the class called?

**Comment by RobinSchmidt** [comment]
however, for everyday parameter smoothing work, EngineersFilter may be a bit too expensive

**Comment by RobinSchmidt** [comment]
rsEngineersFilter - in the filters folder ...yay - i already managed to add the rs prefix there

**Comment by elanhickler** [comment]
you make it super difficult to find where the MODE enum is.................... what is it for lowpass?

**Comment by elanhickler** [comment]
oh wait this time you didn't make it difficult, you actually say InfiniteImpulseResponseDesigner::modes in the comments.

**Comment by elanhickler** [comment]
I wonder if a decorator design would work for engineer's filter with all these modes and types.

myFilter = Bessel(Lowpass(Filter));
myFilter.setCutoff(v);
myFilter.getSample(v);

**Comment by elanhickler** [comment]
except... you'll have a namespace problem.

rosic::FilterType::Bessel(rosic::FilterShape::Lowpass(rosic::FilterBase))

**Comment by elanhickler** [comment]
What is the difference between order and numStages?

**Comment by RobinSchmidt** [comment]
i consider the decorator pattern as a convenient way to mix-and-match additional functionalities to a core class and let client code decide which combination of additional features is needed. in subclassing, you could add one feature per subclass like: MyClass < MyClassWithX < MyClassWithXAndY. but you could never get MyClassWithY (without X) unless you make a separate MyClassWithY class. ...so the number of classes you would have to write would explode with the number of independent features, you'd like to add

**Comment by RobinSchmidt** [comment]
> What is the difference between order and numStages?

numStages is the number of biquad stages. each biquad stage has order 2...or maybe 1 if only 1 pole of the biquad is used

**Comment by elanhickler** [comment]
what do I set? order? and leave numstages alone? im confused!

**Comment by RobinSchmidt** [comment]
yes, order. ignore the BiquadCascade baseclass for setup. use the EngineersFilter methods only. but you need getSampleDirect1 (or 2)  from the baseclass for producing samples

**Comment by RobinSchmidt** [comment]
the numStages will the selected automatically, according to your desired order (and whether it's a lowpass, or bandpass or whatever.

oh, well you should use setPrototypeOrder

**Comment by RobinSchmidt** [comment]
...maybe i should not publically inherit from BiquadCascade. most of it's methods should not be used in an EngineersFilter object

**Comment by RobinSchmidt** [comment]
...but as said...this should be only used for some preliminary experimentation. it will be very wasteful to use full-scale EngineersFilter objects for a task as benign as parameter smoothing

**Comment by elanhickler** [comment]
you don't need a lot at the same time if you have a smoother manager, you only smooth the controls that the user is changing... that's like...2 at most unless you have really long smoothing times.

**Comment by RobinSchmidt** [comment]
ahhh...i see...you would not let the parameter-object itself be responsible for smoothing, but kind of attach a smoother object to it, whenever needed. you could keep a smoother-pool around (in the smoother manager) and the re-use the smoothers from there for any parameter that currently needs smoothing. and in the event you run out of smoothers, the mamanger could dynamically create new ones....interesting. i need to think about that design idea

**Comment by elanhickler** [comment]
uhhhhhhhhh I wrote the code for you already.

**Comment by RobinSchmidt** [comment]
btw. i just discovered this nice pattern site a few days ago. it has a nice collection of design pattern descriptions

**Comment by elanhickler** [comment]
class SmootherManager
{
public:
    SmootherManager() {}
    juce::Array<smoothedParameter *> CurSmoothingList;
    void add(smoothedParameter * sp)
    {
        CurSmoothingList.addIfNotAlreadyThere(sp);
    }

    void remove(smoothedParameter * sp)
    {
        CurSmoothingList.removeFirstMatchingValue(sp);
    }

    void doSmoothing()
    {
        for (auto & sp : CurSmoothingList)
            sp->incValue();
    }
};

the object adds itself to the manager, so you need to instantiate the manager and give it a pointer to that so it can add/remove itself.

add itself when value changed
remove itself when value stops changing, or the difference is like... 0.000001. in that case, set the smoother's internal state to the target value, don't leave it at a difference obviously.

**Comment by RobinSchmidt** [comment]
yes, sounds like a good design. moreover, we could dynamically attach different kinds of smoothers if they have a common baseclass (exponential, gaussian, linear, whatever)

**Comment by RobinSchmidt** [comment]
class Smoother, subclasses: ExponentialSmoother, GaussianSmoother etc. the SmoothingManager would hold an array of pointers to the baseclass. ...or something

**Comment by RobinSchmidt** [comment]
perhaps Smoothers should be applicable to any kind of value (not only Parameters). we could have a class: SmoothableValue and a a Smoother would operate on that. a SmoothableParameter would be subclass of SmoothableValue and Parameter

**Comment by elanhickler** [comment]
whyyyyyyyyyy
![image](https://user-images.githubusercontent.com/9116575/29997460-c0f18e3c-8fc8-11e7-84af-662511dac3ce.png)

**Comment by elanhickler** [comment]
every time I run the debugger, i crash at a different point for engineersfilter

**Comment by elanhickler** [comment]
ok I finally got it to work, I think it will crash randomly, and bessel filter is really nice for smoothing, the only problem is that changing the smoothing amount causes things to get weird... like delays in movement, values jumping around (even if no smoothing is going on).

**Comment by elanhickler** [comment]
for smoothing mouse movements, exponentialsmoother vs bessel filter is like night and day.

**Comment by RobinSchmidt** [comment]
uuhh - random crashes? might this be a threading issue? i think, i will need to check this out in the proper context. changing the smoothing amount? is this the "cutoff-frequency" (i.e. the transition time)?. if so, i'm not surprised, if things go weird. this filter is not really made for "modulating the cutoff frequency", so to speak.

i'd still think, we may want to look into an optimized chain of 1st order lowpasses. this will probably solve the modulation weirdness and make the whole thing more lightweight

**Comment by elanhickler** [comment]
yes, run PrettyScope debugger. Half the time it crashes and doesn't load because of some issue in EngineersFilter code.

**Comment by elanhickler** [comment]
Can you get something working for my mouse smoothing? Really would like to have a better smoother for PrettyScope mouse sooner than later to impress customers.

**Comment by RobinSchmidt** [comment]
wtf is this?
```
smoother.setTargetValue(*this);
//return smoother.getSample();
return smoother2.getSampleDirect1(*this);
```
you are passing a de-referenced this-pointer as function argument? into functions that take a double value?!

**Comment by RobinSchmidt** [comment]
how does that even compile? :-O

**Comment by elanhickler** [comment]
*this converts to jura::parameter::getValue();

did you not read my explanation on how my smoothing system works?

**Comment by RobinSchmidt** [comment]
ahh..ohh. ok, i see. so there's an implicit conversion from the "myparams" object to double due to having defined a double() operator? uuuuhhh.....that's....well.....let's say....non-obvious. i might say confusing. i'm even tempted to say obfuscated. i had initially only a glance at the code because i found it confusing and really trying hard to get my head around it would have been a distraction to what i was doing at the time (the mod-system). i'll take a closer look soon

**Comment by RobinSchmidt** [comment]
but anyway, i think, i could write a smoother class for you based on EngineersFilter which has the same interface as the ExponentialSmoother.

btw. i never had any crahses. i tried to comment out the Exp-smoother stuff and uncomment the engineers-filter stuff. can you give me a version of the code which exposes the crash? i need to see it myself to figure out what's going on

**Comment by elanhickler** [comment]
I just committed a build "this build breaks prettyscope on purpose" with the filter enabled.

but you might run into this build issue: https://github.com/RobinSchmidt/RS-MET/issues/71

**Comment by RobinSchmidt** [comment]
ok - i'll check that tomorrow

**Comment by elanhickler** [comment]
ok, don't use the latest commit, use the commit:

"fixed build issues! yay (Prettyscope still broken on purpose, do not use this commit if your name is not Robin!)"
 daecfea937ecac27fbdd42718f517a184135f4b6

You know how to choose a commit, right?

**Comment by RobinSchmidt** [comment]
yes, i think so. it's not something i use a lot, though.

**Comment by RobinSchmidt** [comment]
ok - i think, i know what the culprit is. you do things like:

```cpp
	paramStrIds =
	{
		&(FXMode = myparams(i++, BUTTON, "FXMode", 0.0, 1.0, 0.0)),
		&(Pause = myparams(i++, BUTTON, "Pause", 0.0, 1.0, 0.0)),
```
where you use the assignment operator = on objects of class type myparams. that implies that such objects are getting copied. when they contain other objects like my EngineersFilter, those get copied too - using the implicitly compiler-generated assignment operator. but the EngineersFilter class contains pointer variables, so the auto-generated trivial assignment operator is not suitable. it creates a shallow copy but we would need a deep copy

**Comment by RobinSchmidt** [comment]
i could manually implement a more suitable assignment operator in EngineersFilter - but actually, i do not intend my dsp objects to be copyable/assignable like variables. things like

`EngineersFilter filter2 = filter1;`

do not seem to make much sense to me. you could instead of having an EngineersFilter member a pointer-to-EngieersFilter and create the object with new in the constructor of myparams and delete it in the destructor (or use a smart-pointer). this way, not the filter itself would get copied but only the pointer-to-it, which should work

**Comment by RobinSchmidt** [comment]
wait...no... i think, using a pointer still wouldn't work

**Comment by elanhickler** [comment]
I got something working, but can't really use engineers filter due to values jumping around when setting smoothing amount. Sometimes the value jumps around and then is permanently offset.

So don't worry about this, will have to wait until you make a new filter class.

**Comment by RobinSchmidt** [comment]
> I got something working

i think, the cleanest solution woud be to get rid of using the = operator here:

`FXMode = myparams(i++, BUTTON, "FXMode", 0.0, 1.0, 0.0)`

by replacing these assignments with something like:

`FXMode. init(i++, BUTTON, "FXMode", 0.0, 1.0, 0.0)`

**Comment by elanhickler** [comment]
ok ill make the change.

so this new gaussian filter you want to make... it could become my favorite filter, bessel filter has no ring. all other filters have ringing. Essentially it's the cleanest possible filter. I'd want to experiment with this filter for general filtering purposes, assuming this new filter will be modulatable.

**Comment by RobinSchmidt** [comment]
i have now declared the copy-constructor and assignment operator deleted in rsEngineersFilter, so now the compiler will balk on any attempt to copy rsEngineersFilter objects. i should really use this idiom consistently throughout the whole library for any class that can't be trivially copied...

i could make a dedicated smoothing filter based on a a chain of 1st order lowpasses. eventually, this would also approach a gaussian shape with increasing order (by virtue of the central limit theorem), but however, it would presumably approach it not as quickly as a real gaussian filter design. so a real gaussian design would be a new design formula to be added to EngineersFilter (i have the formula in a book here). but that wouldn't be modulatable. i guess, i should prefer the 1st order lowpass chain then (i think, it would be modulatable)

**Comment by RobinSchmidt** [comment]
> bessel filter has no ring. all other filters have ringing. Essentially it's the cleanest possible filter. 

the design goal in the bessel filter is to make the phase response as close to linear as possible.

**Comment by elanhickler** [comment]
There's a ton of discussion on the smoothing filter here.

Get started on the upsampling thing for PrettyScope when you can.

**Comment by RobinSchmidt** [comment]
> What I envision is more of an upsampling filter or a realtime bezier system / filter... like... your onepolefilter has an internal sample & hold, I guess instead of using a onepole you would use a system of a history of data points that you interpolate between.

yes - that's what i have in mind. my plan is to record 4 successive samples and put a cubic spline between them such that at the joints (sample-points), the derivatives are matched (i.e. use a "hermite interpolation spline"). and the values of the derivatives (which are inputs to the hermite spline formula) are taken to be differences of adjacent samples. it would introduce a delay/latency of one sample because to find the desired value for the derivative at the "now" sample, i would need one future sample (no issue at all in this case). but: i can straightforwardly implement that spline drawing in the prototype cpu scope - but not in OpenGL. there, the most low-level rendering primitive is already the line (i would set single pixels inside my spline drawing function). but maybe i should do it like that anyway and we may later approximate my spline drawing prototype with (short) lines in OpenGL

**Comment by elanhickler** [comment]
If you can't make it work for PrettyScope it is not useful to me. I thought we would ultimately just add more data points to the audio buffer. Also, that's why I posted the full code of dood.al oscilloscope so you could check how they are doing it.

**Comment by RobinSchmidt** [comment]
> it's not too inefficient to draw only dots btw

that's good. then maybe my original idea of the spline drawing can be used. my algorithm would then tell you where to draw the dots. you would feed it an input sample (pair) and it would spit out a number of dot-coordinates to draw (the number may vary from sample to sample)

**Comment by RobinSchmidt** [comment]
i have already put a stub for the relevant function into the codebase. here's the roadmap:
```cpp
template<class TPix, class TWgt, class TCor>
void rsImagePainter<TPix, TWgt, TCor>::drawDottedSpline(TCor x1, TCor x1s, TCor y1, TCor y1s, 
  TCor x2, TCor x2s, TCor y2, TCor y2s, TPix color, TCor density, int maxNumDots, 
  bool scaleByNumDots)
{
  // Not yet implemented. Here is what we would have to do:
  // -compute coeffs of the two polynomials:
  //  x(t) = a0 + a1*x + a2*x^2 + a3*x^3
  //  y(t) = b0 + b1*y + b2*y^2 + b3*y^3
  // -compute the total length of the spline segment (this will be some kind of analytic line 
  //  integral) to be used to scale the brightness of the dots
  // -compute a sequence of t-values at which to evaluate the polynomials and set a dot - these
  //  t-values should be chosen such that the spline segments between successive t-values have
  //  all the same length, t should be in the range 0..1
  // -the rest is conceptually similar to drawDottedLine
}
```
sooo..that's now what to do...

**Comment by RobinSchmidt** [comment]
ok...making some progress. the connections are smooth now ...but wiggly. i think my data points and derivatives are out of sync. will fix that tomorrow...

**Comment by elanhickler** [comment]
If cpu usage is negligible make it non-optional. Also, if it is negligible, can you make the upsampling algorithm do even more upsampling for even nicer looking results? Or is that not how it works? (like how oversampling can be switched between x2 x4 x8).

Another question, would you be able to use this upsampling inside a synth audio path and get less aliasing? There would be no Gibbs (ripples) like in the antialiasing filter, which is neither good or bad, I would still think to combine it with an antialiasing filter... But the upsampling would perhaps create a smooth (no Gibbs) signal similar to polyblep oscillators but on an arbitrary signal.

**Comment by RobinSchmidt** [comment]
yeah..i did not yet measure cpu usage. and i'm not finished yet because i discovered another problem not apparent above (the plots above use 80 datapoints for one single roundtrip through the lissajous figure). reducing it to 35 datapoints gives this:
![image](https://user-images.githubusercontent.com/1833099/45327699-e182df00-b558-11e8-8f7f-7ff77791a36a.png)
![image](https://user-images.githubusercontent.com/1833099/45327712-eba4dd80-b558-11e8-85f5-7c1616a552b5.png)
...the shape is still nicely smooth, but i get color discontinuities at the joints. i could trace this to having non-equidistant dots along the splines and having high-density segments join to low-density segments. i'm currently pondering how to remedy this.

>  can you make the upsampling algorithm do even more upsampling for even nicer looking results? Or is that not how it works?

no, that's not how it works. i'm just connecting incoming datapoints with spline segments which are conceptually continuous between the datapoints.

> would you be able to use this upsampling inside a synth audio path and get less aliasing? There would be no Gibbs (ripples) like in the antialiasing filter

this should indeed be possible

**Comment by RobinSchmidt** [comment]
the math has become a bit nasty. in order to get the total length of the spline arc, i have to evaluate this integral (in the cyan box):

http://tutorial.math.lamar.edu/Classes/CalcII/ParaArcLength.aspx

from t=0 to t=1 (t is my spline parameter), where the term under the square root boils down to a quartic polynomial. the analytic formula for this integral is too complicated for wolfram:

http://www.wolframalpha.com/input/?i=integral+sqrt(a_0+%2B+a_1+t+%2B+a_2+t%5E2+%2B+a_3+t%5E3+%2B+a_4+t%5E4)+dt+from+0+to+1

i could find analytic formulas for quadratic polynomials under the square root - but even these formulas were a mess. ...so i have to resort to numerical integration. moreover, i don't only need the total arc length but also the inverse function of the function which is defined by this integral (by varying the upper integration limit "beta" keeping alpha=0)......but i'm almost there. i've implemented a nice numerical integration function...now i need to invert the resulting function...almost there...this is fun! i love these kinds of problems.

**Comment by elanhickler** [comment]
oh god kill me.

**Comment by RobinSchmidt** [comment]
it just makes it a bit complicated for me to compute the positions at which we should draw/accumulate dots. with a line, we would just take the total distance between two points (i'll now call the incoming data "points" and the to-be-drawn dots "dots"), get coeffs for a line equation between the points and draw some number of dots along this line. now, we do the same thing but with a spline (obtain spline-coeffs and put dots along this spline) this is what already works and what the pics show. the thing is that with a line, the dots along the line are naturally equidistant - but with a spline, they are not. naively incrementing the parameter t by a fixed stepsize leads to non-equidistant dots, like this:
![image](https://user-images.githubusercontent.com/1833099/45467154-59e6cd00-b71e-11e8-890a-0fe72e389d57.png)
which may lead to these color mismatches at the joints of segments. i'm trying to figure out how to increment my parameter t in unequal amounts so as to have equidistant dots. that's the inversion of the integral....you don't have to worry about this - my function will just spit out the dot-positions for you

**Comment by RobinSchmidt** [comment]
looks like i have the math working. that was a fun challenge. the dots are now equidistant:

![image](https://user-images.githubusercontent.com/1833099/45490217-d30a1280-b766-11e8-94c1-3004cc926993.png)

now, i have a lot of cleaning up, refactoring and optimizing to do

**Comment by elanhickler** [comment]
uhhhhhhhh would it be easy to enable in the phase scope so I can see how it looks before you do the clean up?

**Comment by RobinSchmidt** [comment]
ok - next to the AntiAlias button there's now a new menu with two entries for the interpolation mode: "Linear" (as before) and "Cubic". currently, the cubic mode does not yet use the density compensation. i'll probably make that optional

**Comment by elanhickler** [comment]
cannot open file graphics/realtimespline

rebuilding the toolchain visual studio with jucer didn't work.

**Comment by elanhickler** [comment]
oh because you just added an include and you didn't actually add the file, you must be intending to but didn't do it yet. Commenting out the include got it working.

**Comment by elanhickler** [comment]
### dood.al upsampler
![image](https://user-images.githubusercontent.com/9116575/45539552-e779f980-b7be-11e8-9357-c21d61d41fc1.png)
### your result
some differences can be seen like the O shape at the upper right. The proof seems to be more chaotic looking, not as clean as yours.
![image](https://user-images.githubusercontent.com/9116575/45539453-aeda2000-b7be-11e8-80cc-3d6e509064a3.png)
### proof
![image](https://user-images.githubusercontent.com/9116575/45539468-ba2d4b80-b7be-11e8-8cb2-b09dc12d602f.png)
### ugly!!!!
![image](https://user-images.githubusercontent.com/9116575/45539490-c6b1a400-b7be-11e8-8fb8-800c10a16e67.png)

**Comment by elanhickler** [comment]
### yours / dood.al / proof
![image](https://user-images.githubusercontent.com/9116575/45540595-b3540800-b7c1-11e8-9a7a-ef3b3fd8f04e.png) ![image](https://user-images.githubusercontent.com/9116575/45540614-be0e9d00-b7c1-11e8-99ba-6031891f401d.png) ![image](https://user-images.githubusercontent.com/9116575/45540664-e0081f80-b7c1-11e8-9a0a-60481b596e70.png)

Seems like your upsampling is more accurate, less wiggles.

**Comment by RobinSchmidt** [comment]
> you didn't actually add the file, you must be intending to but didn't do it yet.

oh - right. fixed that. i just added it to the repo. that's where i want to factor out all the code that is related to the 2D spline interpolation which is now scattered over two classes

what do you mean by "proof" and "perfect"? and what is the right image in the last post? linear interpolation?

**Comment by elanhickler** [comment]
proof is the perfect signal using dots to accumulate the correct shape.

**Comment by RobinSchmidt** [comment]
how do you create a perfect signal? using a very low frequency input? my linear and cubic modes also use dots

**Comment by RobinSchmidt** [comment]
ah - ok - i see. but i think, such an approach - a sort of stroboscopic sampling: 
https://www.youtube.com/watch?v=-fSqFWcb4rE
works right only if your input signal is periodic, but your pics look like you are using chaotic signals - which, if i'm not mistaken, should just give you some sort of "cloud" of (pseudo)random points in the areas of interest?

**Comment by elanhickler** [comment]
for the examples, it gives you a good idea of what the lines should be doing since the frequency is low enough for the samplerate to capture enough info.

**Comment by RobinSchmidt** [comment]
i like the single cycle through a  lissajous figure as test. we all know, how the perfect shape should look like (of course, i could also just render a perfect shape by using a high number of samples for the cycle). here, i go down to 15 points per cycle:
![image](https://user-images.githubusercontent.com/1833099/45582126-cabaf000-b8aa-11e8-8b80-65fa7d31e4ad.png)
![image](https://user-images.githubusercontent.com/1833099/45582128-d0b0d100-b8aa-11e8-8546-62672f514f61.png)
for reference always the comparison with linear-interpolsation rendering. with 15 it already starts to look sort of skewed - but still much closer to the true shape than the linear version. i think, for the variation that i have in mind and that i want to try, i will use this shape (and other lissajous figures) as benchmark

**Comment by RobinSchmidt** [comment]
by the way - the dood's scope seems to create a sort of halo around the actual signal:
https://www.youtube.com/watch?v=SvCfz-TY6Go
do you think, that's desirable? is that what analog scopes do?

**Comment by elanhickler** [comment]
Everything is desirable. I would want PrettyScope to be able to look and perform all sorts of interesting and weird ways.

https://youtu.be/wDkG1CgREaQ?t=152
2:32 you see a beautiful persistent image effect

https://youtu.be/mndaenaVClc
here's an oscilloscope exhibiting extreme halo effect

an oscilloscope in slow motion is very beautiful
https://youtu.be/XOAsmd-FFb0

**Comment by RobinSchmidt** [comment]
> an oscilloscope in slow motion is very beautiful

reminds a bit of lightning

aha - so the halo effect is indeed a feature of analog oscilloscopes. ....i remember that we were talking about including a blur effect. i think, that should create a sort of halo. for every frame, instead of just applying pixel-decay, we could additionally apply pixel-blur

**Comment by elanhickler** [comment]
you mean like this: https://www.youtube.com/watch?v=7kI1d7DMbco

**Comment by RobinSchmidt** [comment]
ok - i think, i'm more or less done with the refactoring. it's all in class rsRealTimeSpline now. you set up some buffers for the x,y-coordinates and brightness values, pass them to the object via `setDotBuffers` and for each sample, you call `updateDotBuffers` which fills the dot-buffers with the dots to be drwan for the new segment. should be easy to integrate into prettyscope.

i seem to have introduced a bug into my scope though - in 1D mode, there's a horizontal line of additional dots drawn. will try to sort that out and then on to the next task....maybe modulation indication? or harmonic analysis for ots?

**Comment by elanhickler** [comment]
harmonic analysis for OTS

**Comment by RobinSchmidt** [comment]
i just added a qudratic mode. not sure, how useful this is (may also still be buggy - i have to investigate) and needs to be optimized. now that we have the general framework in place, we may experiment a bit with various other 2D interpolation formulas (i have already some ideas for quartic and quintic formulas in mind)

**Comment by AltTheAi** [comment]
I have no idea how to run this, I have been looking on github for native version of dood.al/oscilloscope and this is the closest. How hard would it be to make this into a standalone project, implement the noise.jpg and drawGrid function and call it a day? What code in this repository do I need to get the line rendering seen in the screenshots here?

**Comment by elanhickler** [comment]
you will want to look at this for dood.al oscilloscope implementation

https://github.com/jameshball/osci-render
https://osci-render.com/

this thread discusses a scope found in ToolChain http://www.rs-met.com/freebies.html and not at all close to dood.al, very different.
